web/webkit/src/main/scala/net/liftweb/http/LiftSession.scala

```diff
@@ -364,7 +364,26 @@ class LiftSession(private[http] val _contextPath: String, val underlyingId: Stri
 
   private val fullPageLoad = new BooleanThreadGlobal
 
-  private val nmessageCallback: ConcurrentHashMap[String, S.AFuncHolder] = new ConcurrentHashMap
+  // Returns a view of the functions stored in the http session
+  private def viewHttpSessionFns:java.util.Map[String, S.AFuncHolder] = java.util.Collections.unmodifiableMap(
+    if(LiftRules.putAjaxFnsInContainerSession) (for {
+        hs <- httpSession
+        map <- Box.legacyNullTest(hs.attribute("ajaxFns").asInstanceOf[ConcurrentHashMap[String, S.AFuncHolder]])
+      } yield map).openOr(new java.util.HashMap[String, S.AFuncHolder])
+    else new java.util.HashMap[String, S.AFuncHolder]
+  )
+
+  // Mutates the http session function map with the given argument and updates the container
+  private def updateHttpSessionFns(f:ConcurrentHashMap[String, S.AFuncHolder]=>Unit):Unit =
+    if(LiftRules.putAjaxFnsInContainerSession) httpSession.foreach { hs =>
+      val map = Box.legacyNullTest(hs.attribute("ajaxFns").asInstanceOf[ConcurrentHashMap[String, S.AFuncHolder]]).openOr(new ConcurrentHashMap[String, S.AFuncHolder])
+      f(map)
+      // Now that the map has been mutated, must write to the session so the container can replicate
+      hs.setAttribute("ajaxFns", map)
+    }
```

Comment:
Though notably that comprehension has different behavior, since the current code will create a map if it can't find it, while the comprehension will simply not run if there's no existing entry.